### PR TITLE
Add thread_started_callback and thread_terminated_callback

### PR DESCRIPTION
**Pull request by n0lavar**
This pull request is adding thread_started_callback and thread_terminated_callback which are called when any internal concurrencpp thread is created. This feature is useful in several cases:
* Profiling. If you want to profile not only the main thread, but all async threads, you should register them in construction and unregister in destruction. This is very useful in gamedev. For example, I was able to do the following with Optick profiler:
image
* Custom general purpose allocator. For example, rpmalloc requires thread registration and unregistration as well

### CODE REVIEW

**Comment by David-Haim** [comment]
Hi there.
I accept your idea.
However, you also need to write tests to this new feature :)
also, If we add callbacks for thread start and thread end we also want the callbacks to accept `std::thread::id` and not just the thread name.

Add those two, then we can continue cooking this pr.
Thanks!

**Comment by n0lavar** [comment]
I think user is always able to use [std::this_thread::get_id](https://en.cppreference.com/w/cpp/thread/get_id) and thus this arg is pretty useless =)

**Comment by David-Haim** [comment]
ok, I also accept that,
but the tests are crucial :)

**Comment by n0lavar** [comment]
Yep
I'll be back with tests

**Comment by David-Haim** [comment]
Hi there,
what is the status of this?

**Comment by n0lavar** [comment]
Hello
Sorry, I didn't have much time during the holidays, but I will definitely finish this request

**Comment by David-Haim** [comment]
Hi there,
I'm working on releasing version 0.1.6, hopefully it'll be out in the next few days.
Unfortunately this will go to the next version.
I will look at this PR in the next weeks (I'm traveling abroad next week, so I won't be available in the next 10 days or so)

**Comment by n0lavar** [comment]
Ok
Please also think about the following:
Don't you want to use string_view instead of char pointer in these callbacks? I used my implementation just because it's easier for me, but it may be better to use modern cpp features.
Do you have any plans to add more arguments to constructor of any of your executors? These args may become more important than callbacks and inserting them before callbacks may break API for lib users.

**Comment by David-Haim** [comment]
> Don't you want to use string_view instead of char pointer in these callbacks? 

That's a good idea!

> Do you have any plans to add more arguments to constructor of any of your executors? 

Currently no, I think the library is API stable now, no major change is planned.

**Comment by David-Haim** [comment]
Please rebase :)

**Comment by n0lavar** [comment]
Thank you for waiting

File: include/concurrencpp/executors/thread_executor.h [hunk 1]
```diff
@@ -20,12 +20,15 @@ namespace concurrencpp {
         std::list<details::thread> m_last_retired;
         bool m_abort;
         std::atomic_bool m_atomic_abort;
+        std::function<void(const char* thread_name)> m_thread_started_callback;
```
Comment (David-Haim) [inline include/concurrencpp/executors/thread_executor.h]: let's go with `std::string_view` like we discussed

File: include/concurrencpp/executors/thread_executor.h [hunk 2]
```diff
@@ -20,12 +20,15 @@ namespace concurrencpp {
...
+        std::function<void(const char* thread_name)> m_thread_terminated_callback;
 
         void enqueue_impl(std::unique_lock<std::mutex>& lock, task& task);
         void retire_worker(std::list<details::thread>::iterator it);
 
        public:
-        thread_executor();
+        thread_executor(const std::function<void(const char* thread_name)>& thread_started_callback = nullptr,
```
Comment (David-Haim) [inline include/concurrencpp/executors/thread_executor.h]: use `{}` instead of `nullptr`

File: include/concurrencpp/executors/thread_pool_executor.h [hunk 3]
```diff
@@ -57,7 +57,11 @@ namespace concurrencpp {
         details::thread_pool_worker& worker_at(size_t index) noexcept;
 
        public:
-        thread_pool_executor(std::string_view pool_name, size_t pool_size, std::chrono::milliseconds max_idle_time);
+        thread_pool_executor(std::string_view pool_name,
+                             size_t pool_size,
+                             std::chrono::milliseconds max_idle_time,
```
Comment (David-Haim) [inline include/concurrencpp/executors/thread_pool_executor.h:62]: as above

File: include/concurrencpp/runtime/runtime.h [hunk 4]
```diff
@@ -33,6 +34,9 @@ namespace concurrencpp {
 
         std::chrono::milliseconds max_timer_queue_waiting_time;
 
+        std::function<void(const char* thread_name)> thread_started_callback;
```
Comment (David-Haim) [inline include/concurrencpp/runtime/runtime.h]: as above

File: include/concurrencpp/threads/thread.h [hunk 5]
```diff
@@ -19,10 +20,23 @@ namespace concurrencpp::details {
...
+               std::function<void(const char* thread_name)> thread_started_callback,
+               std::function<void(const char* thread_name)> thread_terminated_callback) {
+            m_thread = std::thread([name = std::move(name),
+                                    callable = std::forward<callable_type>(callable),
+                                    thread_started_callback = std::move(thread_started_callback),
+                                    thread_terminated_callback = std::move(thread_terminated_callback)]() mutable {
                 set_name(name);
+
+                if (thread_started_callback)
+                    thread_started_callback(name.c_str());
```
Comment (David-Haim) [inline include/concurrencpp/threads/thread.h]: add `{  }` around the `if`.

File: include/concurrencpp/threads/thread.h [hunk 6]
```diff
@@ -19,10 +20,23 @@ namespace concurrencpp::details {
...
+
                 callable();
+
+                if (thread_terminated_callback)
```
Comment (David-Haim) [inline include/concurrencpp/threads/thread.h]: as above

File: source/executors/thread_pool_executor.cpp [hunk 7]
```diff
@@ -57,7 +59,12 @@ namespace concurrencpp::details {
         void ensure_worker_active(bool first_enqueuer, std::unique_lock<std::mutex>& lock);
 
        public:
-        thread_pool_worker(thread_pool_executor& parent_pool, size_t index, size_t pool_size, std::chrono::milliseconds max_idle_time);
+        thread_pool_worker(thread_pool_executor& parent_pool,
+                           size_t index,
+                           size_t pool_size,
+                           std::chrono::milliseconds max_idle_time,
+                           const std::function<void(const char* thread_name)>& thread_started_callback,
```
Comment (David-Haim) [inline source/executors/thread_pool_executor.cpp]: Store these callbacks in the parent threadpool. access them with `m_parent_pool`, no need to store them in the worker itself

```cpp
thread = details::thread(... , m_parent.thread_started_cb, m_parent.thread_terminated_cb);
```

File: test/include/utils/test_thread_callbacks.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
+#ifndef CONCURRENCPP_TEST_THREAD_CALLBACKS_H
+#define CONCURRENCPP_TEST_THREAD_CALLBACKS_H
+
+#include "infra/assertions.h"
+#include "concurrencpp/executors/executor.h"
+
+namespace concurrencpp::tests {
+    template<class executor_fabric_type>
+    void test_thread_callbacks(executor_fabric_type executor_fabric, std::string_view expected_thread_name) {
+        size_t thread_started_callback_invocations_num = 0;
```
Comment (David-Haim) [inline test/include/utils/test_thread_callbacks.h]: these two should be `std::atomic_size_t` otherwise it's a data race.

File: test/include/utils/test_thread_callbacks.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
+#ifndef CONCURRENCPP_TEST_THREAD_CALLBACKS_H
+#define CONCURRENCPP_TEST_THREAD_CALLBACKS_H
+
+#include "infra/assertions.h"
+#include "concurrencpp/executors/executor.h"
+
+namespace concurrencpp::tests {
+    template<class executor_fabric_type>
```
Comment (David-Haim) [inline test/include/utils/test_thread_callbacks.h]: I think that `executor_factory_type` is a clearer name

File: test/source/tests/runtime_tests.cpp [hunk 10]
```diff
@@ -40,6 +40,17 @@ void concurrencpp::tests::test_runtime_constructor() {
     opts.max_background_threads = 7;
     opts.max_background_executor_waiting_time = std::chrono::milliseconds(54321);
 
+    size_t thread_started_callback_invocations_num = 0;
```
Comment (David-Haim) [inline test/source/tests/runtime_tests.cpp]: std::atomic_size_t

File: test/source/tests/runtime_tests.cpp [hunk 11]
```diff
@@ -58,6 +69,30 @@ void concurrencpp::tests::test_runtime_constructor() {
...
+                      &thread_terminated_callback_invocations_num,
+                      start_thread_started_callback_invocations_num,
+                      start_thread_terminated_callback_invocations_num]() {
+                assert_equal(thread_started_callback_invocations_num, start_thread_started_callback_invocations_num + 1);
+                assert_equal(thread_terminated_callback_invocations_num, start_thread_terminated_callback_invocations_num);
+            })
+            .get();
+
+        executor->shutdown();
+        assert_equal(thread_started_callback_invocations_num, start_thread_started_callback_invocations_num + 1);
```
Comment (David-Haim) [inline test/source/tests/runtime_tests.cpp]: I would also test that the thread name is not empty.

**Review by David-Haim — CHANGES_REQUESTED** [review]

File: include/concurrencpp/executors/thread_pool_executor.h (hunk 3 above)
Comment (n0lavar) [reply include/concurrencpp/executors/thread_pool_executor.h:62]: please explain

Comment (David-Haim) [reply include/concurrencpp/executors/thread_pool_executor.h:62]: my mistake, resolved

File: include/concurrencpp/threads/thread.h [hunk 12]
```diff
@@ -19,10 +20,25 @@ namespace concurrencpp::details {
...
+               std::function<void(std::string_view thread_name)> thread_started_callback,
+               std::function<void(std::string_view thread_name)> thread_terminated_callback) {
+            m_thread = std::thread([name = std::move(name),
+                                    callable = std::forward<callable_type>(callable),
+                                    thread_started_callback = std::move(thread_started_callback),
+                                    thread_terminated_callback = std::move(thread_terminated_callback)]() mutable {
                 set_name(name);
+
+                if (thread_started_callback) {
+                    thread_started_callback(name.c_str());
```
Comment (David-Haim) [inline include/concurrencpp/threads/thread.h]: do we actually need `.c_str()` now that everything is `std::string_view` ? I think we can just pass `name`

File: include/concurrencpp/threads/thread.h [hunk 13]
```diff
@@ -19,10 +20,25 @@ namespace concurrencpp::details {
...
+                }
+
                 callable();
+
+                if (thread_terminated_callback) {
+                    thread_terminated_callback(name.c_str());
+                }
```
Comment (David-Haim) [inline include/concurrencpp/threads/thread.h:41]: as above (add static_cast, remove .c_str())

File: include/concurrencpp/threads/thread.h [hunk 14]
```diff
@@ -19,10 +20,25 @@ namespace concurrencpp::details {
...
+               callable_type&& callable,
+               std::function<void(std::string_view thread_name)> thread_started_callback,
+               std::function<void(std::string_view thread_name)> thread_terminated_callback) {
+            m_thread = std::thread([name = std::move(name),
+                                    callable = std::forward<callable_type>(callable),
+                                    thread_started_callback = std::move(thread_started_callback),
+                                    thread_terminated_callback = std::move(thread_terminated_callback)]() mutable {
                 set_name(name);
+
+                if (thread_started_callback) {
```
Comment (David-Haim) [inline include/concurrencpp/threads/thread.h]: add `static_cast<bool>()`

File: include/concurrencpp/timers/timer_queue.h [hunk 15]
```diff
@@ -40,6 +40,8 @@ namespace concurrencpp {
         bool m_abort;
         bool m_idle;
         const std::chrono::milliseconds m_max_waiting_time;
+        std::function<void(std::string_view thread_name)> m_thread_started_callback;
```
Comment (David-Haim) [inline include/concurrencpp/timers/timer_queue.h]: let's make these 2 `cost std::function` as they are not going to be re-assigned.

File: include/concurrencpp/executors/thread_executor.h [hunk 16]
```diff
@@ -20,12 +20,15 @@ namespace concurrencpp {
         std::list<details::thread> m_last_retired;
         bool m_abort;
         std::atomic_bool m_atomic_abort;
+        std::function<void(std::string_view thread_name)> m_thread_started_callback;
```
Comment (David-Haim) [inline include/concurrencpp/executors/thread_executor.h]: let's make these 2 `const std::function` as they are not going to be re-assigned.

File: source/executors/thread_pool_executor.cpp [hunk 17]
```diff
@@ -45,6 +45,8 @@ namespace concurrencpp::details {
         bool m_abort;
         std::atomic_bool m_task_found_or_abort;
         thread m_thread;
+        std::function<void(std::string_view thread_name)> m_thread_started_callback;
```
Comment (David-Haim) [inline source/executors/thread_pool_executor.cpp]: as above (make const)

File: test/source/tests/runtime_tests.cpp [hunk 18]
```diff
@@ -58,6 +71,30 @@ void concurrencpp::tests::test_runtime_constructor() {
     assert_equal(runtime.thread_pool_executor()->max_worker_idle_time(), opts.max_thread_pool_executor_waiting_time);
     assert_equal(runtime.background_executor()->max_concurrency_level(), opts.max_background_threads);
     assert_equal(runtime.background_executor()->max_worker_idle_time(), opts.max_background_executor_waiting_time);
+
+    auto test_runtime_executor = [&thread_started_callback_invocations_num,
+                                  &thread_terminated_callback_invocations_num](std::shared_ptr<executor> executor) {
+        size_t start_thread_started_callback_invocations_num = thread_started_callback_invocations_num;
```
Comment (David-Haim) [inline test/source/tests/runtime_tests.cpp]: why do we actually do it instead of just setting  `thread_started_callback_invocations_num`/ `thread_terminated_callback_invocations_num` to `0` in the start of the function? then we can just test against `0`, `1`

**Review by David-Haim — CHANGES_REQUESTED** [review]

File: test/source/tests/runtime_tests.cpp (hunk 11 above)
Comment (David-Haim) [reply test/source/tests/runtime_tests.cpp]: This is still not added

**Comment by David-Haim** [comment]
Hi
good job.
I reviewed the PR again and added a few more comments :)

**Comment by David-Haim** [comment]
Hi,
I want to release a new concucrrencpp version soon that contains GCC support, 
let's finish this pr? It's just a few minor fixes
thanks :)

File: test/source/tests/runtime_tests.cpp (hunk 11 above)
Comment (n0lavar) [reply test/source/tests/runtime_tests.cpp]: lines 47 and 52

**Comment by n0lavar** [comment]
Hi
Sorry for such a long response

File: include/concurrencpp/executors/thread_executor.h (hunk 16 above)
Comment (David-Haim) [reply include/concurrencpp/executors/thread_executor.h]: has this been added? I don't see it..

**Comment by David-Haim** [comment]
Good job! I'm merging this.
